src/SymfonyCache/KernelClient.php

```diff
@@ -0,0 +1,110 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Zend\Diactoros\ServerRequest;
+
+/**
+ * An implementation of HttpAsyncClient that allows direct routing through the
+ * Symfony HttpCache kernel without executing a real HTTP request.
+ * It uses the HttpFoundationFactory and the Zend DiactorosFactory to convert
+ * between PSR-7 requests and responses. Both are optional dependencies of this
+ * package thus existence of the respective classes is checked for in the
+ * constructor of this client. It is only needed if you have a single node
+ * setup of Symfony and serves as kind of a shortcut for easier configuration.
+ * If you use Varnish or have a multiple node Symfony setup, this class is entirely
+ * useless to you and you can happily ignore it.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class KernelClient implements HttpAsyncClient
+{
+    /**
+     * @var HttpCacheAwareKernelInterface
+     */
+    private $kernel;
+
+    /**
+     * @var HttpFoundationFactory
+     */
+    private $httpFoundationFactory;
+
+    /**
+     * @var DiactorosFactory
+     */
+    private $psr7Factory;
+
+    /**
+     * KernelClient constructor.
+     *
+     * @param HttpCacheAwareKernelInterface $kernel
+     */
+    public function __construct(HttpCacheAwareKernelInterface $kernel)
+    {
+        $this->kernel = $kernel;
+
+        if (!class_exists(HttpFoundationFactory::class)) {
+            throw new \RuntimeException('Install symfony/psr-http-message-bridge to use this client.');
+        }
+
+        if (!class_exists(DiactorosFactory::class)) {
+            throw new \RuntimeException('Install zendframework/zend-diactoros to use this client.');
+        }
+
+        $this->httpFoundationFactory = new HttpFoundationFactory();
+        $this->psr7Factory = new DiactorosFactory();
+    }
+
+    /**
+     * Converts a PSR-7 request to a Symfony HttpFoundation request, runs
+     * it through the HttpCache kernel and converts the response back to a PSR-7
+     * response.
+     *
+     * {@inheritdoc}
+     */
+    public function sendAsyncRequest(RequestInterface $request)
+    {
+        $symfonyRequest = Request::createFromGlobals();
+        $symfonyRequest->server->set('REMOTE_ADDR', '127.0.0.1');
+
+        $query = [];
+        $parts = explode('&', $request->getUri()->getQuery());
+        foreach ($parts as $part) {
+            $chunks = explode('=', $part, 2);
+            $query[$chunks[0]] = $chunks[1];
+        }
+
+        $serverRequest = new ServerRequest(
+            $symfonyRequest->server->all(),
+            [],
+            $request->getUri(),
+            $request->getMethod(),
+            $request->getBody(),
+            $request->getHeaders(),
+            $symfonyRequest->cookies->all(),
+            $query
+        );
+
+        $symfonyRequest = $this->httpFoundationFactory->createRequest($serverRequest);
```

Comment:
why first create a PSR server request and then convert it to a symfony request? we do have to convert the request we get from the http dispatcher but i think we could create a symfony server request directly, no?

Reply:
I don't understand this comment.

Reply:
if i am not mistaken, we have a method that gets a psr request, creates a symfony request, creates a diactoros server request and then converts the diactoros server request to a symfony server request that is passed to the handle. or am i misreading this?

Reply:
You are misreading I guess. We get a PSR-7 request and need a Symfony Request.